File: src-tauri/src/models/workflow.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Default)]
pub struct WorkflowContext {
    pub variables: HashMap<String, String>,
}

impl WorkflowContext {
    pub fn get(&self, key: &str) -> Option<&String> {
        self.variables.get(key)
    }

    pub fn set(&mut self, key: String, value: String) {
        self.variables.insert(key, value);
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Workflow {
    pub steps: Vec<Step>,
    pub flow: Flow,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum Step {
    Input(InputStep),
    Information(InformationStep),
    Check(CheckStep),
    Condition(ConditionStep),
    Command(CommandStep),
    Choice(ChoiceStep),
    File(FileStep),
}

//
// INPUT
//

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InputStep {
    pub id: String,
    pub name: String,
    pub variable_name: String,
    pub prompt: String,

    pub description: Option<String>,
    pub default_value: Option<String>,
    pub required: bool,
    pub secret: bool,
    pub validation_regex: Option<String>,
}

//
// INFORMATION
//

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InformationStep {
    pub id: String,
    pub name: String,
    pub title: String,
    pub content: String,
}

//
// CHECK
//

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CheckStep {
    pub id: String,
    pub name: String,
    pub check_type: CheckType,

    pub on_success: Option<Workflow>,
    pub on_failure: Option<Workflow>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum CheckType {
    Command(CommandCheck),
    FileExists(FileExistsCheck),
    DirectoryExists(DirectoryExistsCheck),
    EnvironmentVariable(EnvironmentVariableCheck),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CommandCheck {
    pub id: String,
    pub command: String,
    pub expected_exit_code: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FileExistsCheck {
    pub id: String,
    pub path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DirectoryExistsCheck {
    pub id: String,
    pub path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EnvironmentVariableCheck {
    pub id: String,
    pub variable_name: String,
}

//
// CONDITION
//

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConditionStep {
    pub id: String,
    pub name: String,

    pub left: String,
    pub operator: ConditionOperator,
    pub right: String,

    pub on_true: Option<Workflow>,
    pub on_false: Option<Workflow>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum ConditionOperator {
    Equals,
    NotEquals,
    GreaterThan,
    GreaterThanOrEqual,
    LessThan,
    LessThanOrEqual,
    Contains,
}

//
// COMMAND
//

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CommandStep {
    pub id: String,
    pub name: String,
    pub command: String,

    pub working_directory: Option<String>,

    pub capture_stdout_to: Option<String>,
    pub capture_stderr_to: Option<String>,
    pub capture_exit_code_to: Option<String>,

    pub on_success: Option<Workflow>,
    pub on_failure: Option<Workflow>,
}

//
// CHOICE
//

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChoiceStep {
    pub id: String,
    pub name: String,

    pub variable_name: String,
    pub prompt: String,

    pub options: Vec<String>,

    pub default_value: Option<String>,
    pub allow_custom_value: bool,
}

//
// FILE
//

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FileStep {
    pub id: String,
    pub name: String,

    pub operation: FileOperation,
    pub file_path: String,

    pub on_success: Option<Workflow>,
    pub on_failure: Option<Workflow>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum FileOperation {
    CreateOrOverwrite {
        content: String,
    },

    Append {
        content: String,
    },

    ReplaceText {
        search: String,
        replace: String,
    },
}

//
// JUMP
//

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum Flow {
    Continue,
    Jump {
        target_step_id: String,
    },
}

File: src/models/workflow.ts
export type WorkflowBranch =
  | "onSuccess"
  | "onFailure"
  | "onTrue"
  | "onFalse";

export type WorkflowPath = Array<
  string | WorkflowBranch
>;


//
// FLOW
//

export type FlowType =
  | {
      type: "continue";
    }
  | {
      type: "jump";
      targetStepId: string;
    };

export interface FlowStep extends BaseStep {
  type: "flow";

  flowType: FlowType;
}

//
// WORKFLOW
//

export interface Workflow {
  id: string;
  name: string;
  description?: string;
  environment?: string;
  createdAt: string;
  updatedAt: string;

  steps: Step[];
}

//
// BASE STEP
//

export interface BaseStep {
  id: string;
  name: string;
}

//
// INPUT
//

export interface InputStep extends BaseStep {
  type: "input";

  variableName: string;
  prompt: string;

  description?: string;
  defaultValue?: string;
  placeholder?: string;

  required: boolean;
  secret: boolean;

  validationRegex?: string;
}

//
// INFORMATION
//

export interface InformationStep extends BaseStep {
  type: "information";

  title: string;
  content: string;
}

//
// CHECK
//

export interface CheckStep extends BaseStep {
  type: "check";

  checkType: CheckType;

  onSuccess: Workflow;
  onFailure: Workflow;
}

export type CheckType =
  | CommandCheck
  | FileExistsCheck
  | DirectoryExistsCheck
  | EnvironmentVariableCheck;

export interface CommandCheck {
  type: "command";

  command: string;
  expectedExitCode: number;
}

export interface FileExistsCheck {
  type: "fileExists";

  path: string;
}

export interface DirectoryExistsCheck {
  type: "directoryExists";

  path: string;
}

export interface EnvironmentVariableCheck {
  type: "environmentVariable";

  variableName: string;
}

//
// CONDITION
//

export interface ConditionStep extends BaseStep {
  type: "condition";

  variableName: string;
  operator: ConditionOperator;
  value: string;

  onTrue: Workflow;
  onFalse: Workflow;
}

export type ConditionOperator =
  | "equals"
  | "notEquals"
  | "greaterThan"
  | "greaterThanOrEqual"
  | "lessThan"
  | "lessThanOrEqual"
  | "contains";

//
// COMMAND
//

export interface CommandStep extends BaseStep {
  type: "command";

  command: string;

  workingDirectory?: string;

  captureStdoutTo?: string;
  captureStderrTo?: string;
  captureExitCodeTo?: string;

  onSuccess: Workflow;
  onFailure: Workflow;
}

//
// CHOICE
//

export interface ChoiceStep extends BaseStep {
  type: "choice";

  variableName: string;
  prompt: string;

  options: string[];

  defaultValue?: string;

  allowCustomValue: boolean;
}

//
// FILE
//

export interface FileStep extends BaseStep {
  type: "file";

  filePath: string;

  operation: FileOperation;

  onSuccess: Workflow;
  onFailure: Workflow;
}

export type FileOperation =
  | CreateOrOverwriteOperation
  | AppendOperation
  | ReplaceTextOperation;

export interface CreateOrOverwriteOperation {
  type: "createOrOverwrite";

  content: string;
}

export interface AppendOperation {
  type: "append";

  content: string;
}

export interface ReplaceTextOperation {
  type: "replaceText";

  search: string;
  replace: string;
}

//
// STEP UNION
//

export type Step =
  | InputStep
  | InformationStep
  | CheckStep
  | ConditionStep
  | CommandStep
  | ChoiceStep
  | FileStep
  | FlowStep;

export interface WorkflowSummary {
  id: string;
  name: string;
  description?: string;
  createdAt: string;
  updatedAt: string;
}